Aaq 278 create backend pipeline

yvonnegitau opened the pull request:
Reviewer: Suzin

Estimate: 30 mins

---

## Ticket

Feature:https://idinsight.atlassian.net/browse/AAQ-278

## Description, Motivation and Context

### Goal
The goal of the task is to create a workflow for the backend. This workflow will deploy the backend to ECS when merged to the demo branch.

### Changes
The changes made include;

1. Creating a deployment folder under core_backend folder. The folder will consist of the docker_compose file and bootstrap_{env}.sh file. The bootstrap file is used to initialise secrets stored in AWS Secret manager.
2. `deploy_backend.yaml` file. This workflow will be triggered only when there is a change in the file or the `core_backend` folder.
3. Environment variables were added to Github. If a new branch is added, a new Environment will be created and the variables will be added. There will be no change to the workflow unless a step is to be added.

## How has this been tested?
This was used to deploy the current demo application.

## To-do before merge

- [x] Ensure PR #65 is merged

## Checklist

Fill with `x` for completed. Delete any lines that are not relevant

- [ ] My code follows the style guidelines of this project
- [ ] I have reviewed my own code to ensure good quality
- [ ] I have tested the functionality of my code to ensure it works as intended
- [ ] I have resolved merge conflicts
- [ ] I have updated the automated tests (if applicable)
- [ ] I have updated the requirements (if applicable)
- [ ] I have updated the README file (if applicable)
- [ ] I have updated affected documentation (if applicable)

suzinyou commented on .gitignore:
```diff
@@ -1,7 +1,7 @@
-.env*
-**/data/certbot/
-**/data/nginx/app.conf
+.env*	
+**/data/certbot/	
+**/data/nginx/app.conf	
```
Why does this keep happening 😆 
```suggestion
.env*
**/data/certbot/
**/data/nginx/app.conf
```

suzinyou reviewed (commented):
Leaving some initial suggestions!

suzinyou commented:
How would you recommend I test it?

yvonnegitau commented:
> How would you recommend I test it?

I followed the following steps to test:

1. created an .env file and place all the variables there. 
2. changed the push condition to be name of the branch instead of demo
3. replaced the `${{github.ref_name}}` with demo
4. Instead of `{{}secrets.<env_value>}` i used `{{env.<env_value>}}` since I was getting the values from a file.
The alternative is to create the demo environment and push the code to demo branch and it will deploy the code

yvonnegitau commented on admin_app/deployment/docker-compose.yml:
```diff
@@ -0,0 +1,16 @@
+version: "3"
+services:
+  frontend-container:
+    image: ${ACCOUNT}.dkr.ecr.${AWS_REGION}.amazonaws.com/aaq-demo-ecr-repository:frontend
```
This will be changed in Admin app PR. For backend I have added in code comments. We can also add this to the README

yvonnegitau commented on .github/workflows/task-definition.json:
This will temporarily hold the task definition when it sis downloaded.

yvonnegitau commented on .github/workflows/deploy_backend.yml:
```diff
@@ -0,0 +1,146 @@
+                ecs-cli configure --cluster ${{ secrets.CLUSTER_NAME }} \
+                --region ${{ secrets.AWS_REGION }} \
+                --default-launch-type EC2 \
+                --config-name aaq-demo-backend-config
```
The config name is not tied to anything. You can use any name for this. We can have a stated naming convention to follow

yvonnegitau commented on .github/workflows/deploy_backend.yml:
```diff
@@ -0,0 +1,146 @@
+              with:
+                context: core_backend
+                platforms: linux/arm64
+                file: core_backend/Dockerfile.demo
```
The difference is the entrypoint. Until we fully move from dev, we will need the two

yvonnegitau commented on core_backend/deployment/bootstrap_demo.sh:
```diff
@@ -23,23 +23,31 @@ function get_secret_value() {
+SECRET_WHATSAPP="${PROJECT_NAME}-${ENV}-whatsapp-token"
+SECRET_OPENAI="${PROJECT_NAME}-${ENV}-open-ai-key"
+SECRET_QUESTION_ANSWER="${PROJECT_NAME}-${ENV}-question-answer"
+
```
Made changes to the secrets in bootstrap. Thought that this would be the default bootstrap then if we need to make specific changes to a project or environment, we can create a new bootstrap

suzinyou commented on .github/workflows/deploy_backend.yml:
```diff
@@ -5,144 +5,142 @@ on:
+          AWS_ACCOUNT_ID: ${{ secrets.AWS_ACCOUNT_ID }}
+          AWS_REGION: ${{ secrets.AWS_REGION }}
+          QDRANT_HOST: ${{ secrets.QDRANT_HOST }}
+          NEXT_PUBLIC_BACKEND_URL: ${{ secrets.NEXT_PUBLIC_BACKEND_URL }}
```
Fixed `NEXT_PUBLIC_API_URL` to `NEXT_PUBLIC_BACKEND_URL`

suzinyou commented on infrastructure/demo/network/vpc_endpoints.tf:
```diff
@@ -81,5 +81,15 @@ resource "aws_vpc_endpoint_policy" "vpc_endpoint_s3_access" {
           "arn:aws:s3:::amazonlinux-2-repos-${var.region}/*"
         ]
       },
+      {
+        Effect    = "Allow",
+        Principal = "*",
+        Action = [
+          "s3:getObject",
+        ],
+        Resource = [
+          "arn:aws:s3:::prod-${var.region}-starport-layer-bucket/*"
+        ]
+      },
```
Additional permission required for ECS task role to pull image from ECR

suzinyou commented on requirements-dev.txt:
```diff
@@ -2,7 +2,7 @@ ruff==0.0.289
 black==23.9.1
 ruff-lsp==0.0.39
-mypy==1.5.1
+mypy==1.8.0
```
Upgrade mypy (see : https://github.com/encode/starlette/discussions/2423)
